contributors/devel/sig-scheduling/scheduling_code_hierarchy_overview.md

```diff
@@ -0,0 +1,280 @@
+# Scheduler code hierarchy overview
+
+## Introduction
+
+The scheduler watches for newly created Pods that have no Node assigned.
+For every Pod that the scheduler discovers, the scheduler becomes responsible
+for finding the best Node for that Pod to run on.
+Scheduling in general is quite an extensive field in computer science which takes
+into account various range of constraints and limitations.
+Each workload may require a different approach to achieve optimal scheduling results.
+The kube-scheduler provided by Kubernetes project was constructed with a goal
+to provide high throughput at the cost of being simple.
+To help in building a scheduler (the default or a custom one) and to share
+elements of the scheduling logic,
+[the scheduling framework](https://www.google.com/url?q=https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/&sa=D&source=editors&ust=1617027676318000&usg=AOvVaw2vEsCZwet-BiQGhy0mVpZu)
+was implemented.
+The framework does not provide all pieces to build a new scheduler from scratch.
+Queues, caches, scheduling algorithms and other building elements are still needed to assemble
+a fully functional unit. This document aims at describing how all the individual
+pieces are put together and what’s their role in the overall architecture
+so a developer can quickly orient in the code.
+
+## Scheduling a pod
+
+The default scheduler instance has a loop running indefinitely
+which (everytime there’s a pod) is responsible for invoking the scheduling logic
+and making sure a pod gets either a node assigned or requeued for future processing.
+Each loop consists of a scheduling and a binding cycle.
+The scheduling cycle is responsible for running the scheduling algorithm selecting
+the most suitable node for placing the pod.
+The binding cycle makes sure the kube-apiserver is made aware of the selected
+node at the right time (e.g. gang scheduling requires to wait until all pods
```

Comment:
We may reword it like this:

```
A pod may be bound immediately, or in the case of gang scheduling,
wait until all its sibling pods have their node assigned.
```

We shouldn't word it as "binding all pods at once" as each binding cycle is a goroutine. A single binding goroutine can only cover its own pod.